primefield/src/fiat.rs

```diff
@@ -0,0 +1,239 @@
+//! Macros for generating wrappers for `fiat-crypto` synthesized field implementations.
+
+/// Add `fiat-crypto` synthesized arithmetic impls to the given field element.
+#[macro_export]
+macro_rules! fiat_field_arithmetic {
+    (
+        $fe:tt,
+        $bytes:ty,
+        $uint:ty,
+        $non_mont_type: expr,
+        $mont_type: expr,
+        $from_mont:ident,
+        $to_mont:ident,
+        $add:ident,
+        $sub:ident,
+        $mul:ident,
+        $neg:ident,
+        $square:ident,
+        $divstep_precomp:ident,
+        $divstep:ident,
+        $msat:ident,
+        $selectznz:ident
+    ) => {
+        impl $fe {
+            /// Decode [`
+            #[doc = stringify!($fe)]
+            /// `] from [`
+            #[doc = stringify!($uint)]
+            /// `] converting it into Montgomery form.
+            ///
+            /// Does *not* perform a check that the field element does not overflow the order.
+            ///
+            /// Used incorrectly this can lead to invalid results!
+            #[inline]
+            pub(crate) const fn from_uint_unchecked(w: $uint) -> Self {
+                let mut out = $mont_type([0; <$uint>::LIMBS]);
+                $to_mont(&mut out, &$non_mont_type(w.to_words()));
+                Self(<$uint>::from_words(out.0))
+            }
+
+            /// Translate [`
+            #[doc = stringify!($fe)]
+            /// `] out of the Montgomery domain, returning a [`
+            #[doc = stringify!($uint)]
+            /// `] in canonical form.
+            #[inline]
+            pub const fn to_canonical(self) -> $uint {
+                let mut out = $non_mont_type([0; <$uint>::LIMBS]);
+                $from_mont(&mut out, &$mont_type(self.0.to_words()));
+                <$uint>::from_words(out.0)
+            }
+
+            /// Add elements.
+            #[inline]
+            pub const fn add(&self, rhs: &Self) -> Self {
+                let mut out = $mont_type([0; <$uint>::LIMBS]);
+                $add(
+                    &mut out,
+                    &$mont_type(self.0.to_words()),
+                    &$mont_type(rhs.0.to_words()),
+                );
```

Comment:
The newtypes introduced in newer versions of `fiat-crypto` make these conversions somewhat annoying, and they're occurring wherever we invoke any generated code.

Fortunately sprinkling `#[inline]` on it seems to be sufficient to get the compiler to optimize them away, but it might make sense to support casting between generated field element types (which contain `crypto-bigint::Uint`s) and references to their `fiat-crypto` generated newtypes.

The main thing that's missing right now is the `fiat-crypto` newtypes aren't annotated as `#[repr(transparent)]`, as what's really needed here is a cast between the inner array type and the newtype. That shouldn't be too difficult to add in `fiat-constify` though.